gordon_janitor_gcp/gce_clients.py

```diff
@@ -0,0 +1,297 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Client modules to to retrieve project and instance data from GCE.
+
+These clients use the asynchronous HTTP client defined in
+:py:mod:`gordon_janitor_gcp.http_client` and require service account or
+JWT-token credentials for authentication.
+
+To use:
+
+.. code-block:: python
+
+    import asyncio
+
+    import aiohttp
+
+    from gordon_janitor_gcp import auth
+
+    async def main():
+        session = aiohttp.ClientSession()
+        loop = asyncio.get_event_loop()
+        auth_client = auth.GoogleAuthClient(k
+            eyfile='/path/to/keyfile', session=session)
+        client = GCPClient(auth_client, session)
+        instances = await client.list_instances('project-id')
+        print(instances)
+
+    loop.run_until_complete(main())
+    # example output
+    # [{'hostname': 'instance-1', 'internal_ip': '10.10.10.10',
+    #   'external_ip': '192.168.1.10'}]
+"""
+
+import asyncio
+import logging
+
+from gordon_janitor_gcp import exceptions
+from gordon_janitor_gcp import http_client
+
+
+class PagingGoogleClientBase(http_client.AIOGoogleHTTPClient):
+    """HTTP client base class that extracts data from paginated Google APIs.
+
+    Args:
+        auth_client (gordon_janitor_gcp.auth.GoogleAuthClient): client
+            to manage authentication for HTTP API requests.
+        session (aiohttp.ClientSession): (optional) ``aiohttp`` HTTP
+            session to use for sending requests. Defaults to the
+            session object attached to ``auth_client`` if not provided.
+    """
+
+    def __init__(self, auth_client=None, session=None):
+        super().__init__(auth_client=auth_client, session=session)
+
+    def parse_response_into_items(self, response, items):
+        """Parse an API response and append the results to ``items``
+
+        Must be implemented by inheriting class.
+
+        Args:
+            response (dict): API json response.
+            items (list): list for collecting parsed results.
+        """
+        raise NotImplementedError(
+            f'{self.__class__.__name__} implement "parse_response_into_items" '
+            'method!')
+
+    async def list_all_items(self, url, params, retries, retry_wait):
+        """Aggregate data from all pages of an API query.
+
+        Args:
+            url (str): Google API endpoint URL.
+            params (dict): URL query parameters.
+            retries (int): Number of times to retry an API call.
+            retry_wait (int): Number of seconds to wait between each retry.
+        Returns:
+            list: parsed query response results.
+        """
+        items = []
+        next_page_token = None
+
+        while True:
```

Comment:
I've added some basic retry logic here, but it feels like it would simpler to have that logic in the client itself or make use of a 3rd party module. If we decide to use our own logic, we should change this to an exponential backoff and introduce timeouts to the requests since we've been bitten by this before.